core-blocks/file/edit.js

```diff
@@ -0,0 +1,252 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { getBlobByURL, revokeBlobURL } from '@wordpress/utils';
+import {
+	ClipboardButton,
+	IconButton,
+	Toolbar,
+	withNotices,
+} from '@wordpress/components';
+import { Component, compose, Fragment } from '@wordpress/element';
+import {
+	MediaUpload,
+	MediaPlaceholder,
+	BlockControls,
+	RichText,
+	editorMediaUpload,
+} from '@wordpress/editor';
+
+/**
+ * Internal dependencies
+ */
+import './editor.scss';
+import FileBlockInspector from './inspector';
+import FileBlockEditableLink from './editable-link';
+
+class FileEdit extends Component {
+	constructor() {
+		super( ...arguments );
+
+		const {
+			href,
+			showDownloadButton = true,
+			buttonText = __( 'Download' ),
+			id,
+		} = this.props.attributes;
+
+		this.onSelectFile = this.onSelectFile.bind( this );
+
+		// Initialize default values if undefined
+		this.props.setAttributes( {
+			showDownloadButton,
+			buttonText,
+		} );
+
+		// edit component has its own attributes in the state so it can be edited
+		// without setting the actual values outside of the edit UI
+		this.state = {
+			showPlaceholder: ! href,
+			href,
+			attachmentPage: undefined,
+			showCopyConfirmation: false,
+		};
+
+		if ( id !== undefined ) {
+			// Get Attachment Page URL
+			wp.apiRequest( {
+				path: `/wp/v2/media/${ id }`,
+				method: 'GET',
+			} )
+				.then(
+					( result ) => {
+						this.setState( { attachmentPage: result.link } );
```

Comment:
Would the `setState` work correctly if the component hasn't been rendered yet?

Reply:
Good point. Should've been in `componentDidMount()`.